### ISSUE TITLE
handle all session modifying postgres statements like transactions

### ISSUE THREAD

**Issue by tpo** [pgbouncer/pgbouncer#396]
Hello,

TLDR: would it be possible to move a frontend session into its own backend session as soon as a session context modifying statement happens? 

Context:

after a lot of effort we found out, that restoring a pg_dump would break independent postgres sessions. This was due to pg_dumps containing by default a line like:

    SELECT pg_catalog.set_config('search_path', '', false);

... which will break other active psql sessions, since all of a sudden the "public" namespace/schema space will not be considered by SQL queries any more. A simple query like:

    SELECT * from my_table;

... will not find/see "my_table" any more. Only

    SELECT * from public.my_table;

... will work.

This happens in "transaction" pgbouncer pooling mode. I expect it not to happen in "session" pooling mode.

However using "session" pooling mode is not really interesting for us, since it will create a new connection to the backend and thus create a new "postgres" process for each new session. Thus "session" pooling mode will annihilate the effect which we are employing pgbouncer for, namely to reduce the number of backend "postgres" processes and the associated ressources they need.

In order to be able to benefit from "transaction" mode pgbouncer would need to "detect all postgres/SQL statements" that modify session state, that is:

* cursors: https://github.com/yandex/odyssey/issues/45
* prepared statements: https://pgbouncer.github.io/faq.html#how-to-use-prepared-statements-with-session-pooling
* session settings like pg_catalog.set_config
* etc.

... and move a session where a statement would change session state into its own backend connection.
 
In general, most of the things that "DISCARD ALL" resets would need to be moved into their own backend connection.

Can functionality like this be implemented? Us sponsoring the development of such a feature is open to discussion.

?

Thanks,
*t

**Comment by petere** [comment]
By design, PgBouncer does not inspect or analyze the queries that are being sent through. So detecting this would have to be done at the protocol level somehow.

There is [discussion on pgsql-hackers](https://www.postgresql.org/message-id/flat/ac873432-31cf-d5e4-0b80-b5ac95cfe385@postgrespro.ru) about something very similar in the context of a proposed built-in connection pool. They speak of "tainting" a backend when it changes session state. But then it is also discussed that this would trigger too often to make it useful in that simple form.

So this is a very complicated issue, and I don't see a PgBouncer-only solution emerging any time soon.

**Comment by tpo** [comment]
Hi Peter,

On Sat, 3 Aug 2019, Peter Eisentraut wrote:

> By design, PgBouncer does not inspect or analyze the queries that are being sent through. So detecting this would have to be done at the protocol level somehow.
> 
> There is discussion on pgsql-hackers about something very similar in the context of a proposed built-in connection pool. They speak of "tainting" a backend when it changes session state. But then it is also discussed
> that this would trigger too often to make it useful in that simple form.
> 
> So this is a very complicated issue, and I don't see a PgBouncer-only solution emerging any time soon.

Many thanks for keeping my ticket in the back of your head and for the 
pointer. That discussion on pgsql-hackers concerns the problem we are 
having precisely, so I am very much looking forward to how the proposed 
solution will evolve.

You write:

> They speak of "tainting" a backend when it changes session state. But 
> then it is also discussed that this would trigger too often to make it 
> useful in that simple form.

In our case (qgiscloud.com) we have a many readers (SELECT & calculations) 
and very, very few writers situation. Also most of the readers exists in 
a somewhat controlled context, where "tainting" "shoud not" occur.

So I think in our case the tainting problem quite confined and sparse, and 
so the "tainting" approach would in effect solve an important problem for 
us, namely that a very small number of postgres clients do break sessions, 
by changing session state of a multiplexed connection.

So again, *thanks a lot* for sparing some thoughts about this, and for the 
pointer!
*t

**Comment by petere** [comment]
I'm going to close this issue, since a solution if any would not originate in PgBouncer.
